feat: Combobox, MultiInput, Select fiori3

JKMarkowski opened the pull request:
## Related Issue
fixes: https://github.com/SAP/fundamental-styles/issues/596
fixes: https://github.com/SAP/fundamental-styles/issues/572
fixes: https://github.com/SAP/fundamental-styles/issues/573
fixes: https://github.com/SAP/fundamental-styles/issues/604

## Description
In this PR I removed old `dropdown.scss` and `multi-input.scss`. It got replaced by new modifiers for `list` component, which are `--multi-input` and `--dropdown`.

### Dropdown Component usage:
For combobox, it will use `popover`, `input-group` and **`list with --dropdown modifier`** component
Multi Input - now it uses `popover`, `token`, `tokenizer`, `input-group` and **`list with --mult-input modifier`** components
Select has it's own styling, so it uses `popover`, `select` and  **`list with --dropdown modifier`** components

### New featues:
To adapt some components, that can be used along with Combobox, MultiInput and Select, I had to add some modifiers.

**Select**
Select got new class `fd-select__button`. Before it was pseudo element, but we should get rid of  pseudo elements unnecessary usage.

**InputGroup**
`fd-input-group` class got new modifier `fd-input-group--control`, to reacts on `hover`/`active` effects.
`fd-input-group__button` should have `is-expanded` class, when popover is expanded.

**Popover**
To add limitations of width, specified in documentation, there is new modifier for class `fd-popover__body`, it's `fd-popover__body--dropdown`.

**Form Messages**
Before messages were used only inside `popover` component. When it's used with Select, MultiInput, Combobox, it should stay visible all the time, so there is added `fd-form-message--static` modifier.

### Docs
JS functionality on our docs page had to be changed, actions with `aria-controls` for dropdowns will propagate to all elements with the same `aria-controls ID`.
 
### Test Cases
Tests for `dropdown` and `multi-input` are inside `list` component.

### Removed classes
* `fd-dropdown` class and file is completely removed
* `fd-multi-input` class and file is completely removed

## Screenshots
> **NOTE:** If you've made any style changes, please provide appropriate screenshots (before and after) to help reviewers.
>
> To see examples of which screenshots to include, go to [Screenshot Examples](https://github.com/SAP/fundamental-styles/wiki/Pull-Request-Screenshot-Examples).

### Before:

Combobox
![image](https://user-images.githubusercontent.com/26483208/73354879-0b948880-4297-11ea-8b5b-2f307bc9e194.png)

Compact Combobox
![image](https://user-images.githubusercontent.com/26483208/73354893-1818e100-4297-11ea-9923-6aa091605435.png)

MutliInput
![image](https://user-images.githubusercontent.com/26483208/73354909-2404a300-4297-11ea-9143-4cb564205a65.png)

Compact MultiInput
![image](https://user-images.githubusercontent.com/26483208/73354923-2f57ce80-4297-11ea-8cff-6581dc72e1ad.png)


Select
![image](https://user-images.githubusercontent.com/26483208/73354945-3c74bd80-4297-11ea-9dea-31a73ea1cbb1.png)

Compact Select
![image](https://user-images.githubusercontent.com/26483208/73354961-45658f00-4297-11ea-835a-5f55cb64cc15.png)


### After:

Combobox:
![image](https://user-images.githubusercontent.com/26483208/73591400-75f33600-44ee-11ea-9596-3c5c307b0443.png)

Compact Combobox with more text:
![image](https://user-images.githubusercontent.com/26483208/73591404-7be91700-44ee-11ea-9892-5487da099c39.png)

Combobox with Message:
![image](https://user-images.githubusercontent.com/26483208/73591406-86a3ac00-44ee-11ea-8350-bad4660d5e1b.png)

Combobox Disabled:
![image](https://user-images.githubusercontent.com/26483208/73591410-8d322380-44ee-11ea-96a8-9c4963abfc92.png)

combobox readonly:
![image](https://user-images.githubusercontent.com/26483208/73591414-93c09b00-44ee-11ea-86c6-c0305078e849.png)


Multi Input  with more text
![image](https://user-images.githubusercontent.com/26483208/73591416-9ae7a900-44ee-11ea-803f-4c1b72187979.png)

Multi Input Compact:
![image](https://user-images.githubusercontent.com/26483208/73591423-a33fe400-44ee-11ea-898b-676108b32f5b.png)

REadonly Multi Input with Popover
![image](https://user-images.githubusercontent.com/26483208/73591433-acc94c00-44ee-11ea-9ea5-6d515512cd68.png)

Disabled Multi Input
![image](https://user-images.githubusercontent.com/26483208/73591436-b357c380-44ee-11ea-941b-81323ddffaa7.png)

Multi Input with Message
![image](https://user-images.githubusercontent.com/26483208/73591440-ba7ed180-44ee-11ea-9388-789dad4b1bd8.png)

Select:
![image](https://user-images.githubusercontent.com/26483208/73591447-c1a5df80-44ee-11ea-83dd-fbacd690d49f.png)

Compact Select with more text:
![image](https://user-images.githubusercontent.com/26483208/73591448-c79bc080-44ee-11ea-9fe6-e4759c6706c3.png)

Two Columns Select:
![image](https://user-images.githubusercontent.com/26483208/73591453-d1252880-44ee-11ea-9d89-4ad5209e408a.png)

Disabled Select
![image](https://user-images.githubusercontent.com/26483208/73591459-ea2dd980-44ee-11ea-8451-758164070c15.png)

Readonly Select:
![image](https://user-images.githubusercontent.com/26483208/73591458-e4d08f00-44ee-11ea-834a-166ee8c98798.png)


Semantic States with Messages
![image](https://user-images.githubusercontent.com/26483208/73591460-f1ed7e00-44ee-11ea-805a-c7eb555cd1f8.png)

![image](https://user-images.githubusercontent.com/26483208/73591461-f9ad2280-44ee-11ea-815b-a484752f5158.png)

estelaxu3 commented:
Visuals looking good to me.

JKMarkowski commented:
Thanks @estelaxu3, Anyway the PR's code and preview are ready for review.

salarenko commented on src/input-group.scss:
```diff
@@ -201,6 +201,31 @@ $fd-input-border-radius: var(--sapField_BorderCornerRadius);
 
     border: none;
     width: 100%;
+
```
Please update component structure description (Same for the rest of edited components)
```
/*!
.fd-input-group+(--inline)
    .fd-input-group__addon+(--button)
*/
```

JKMarkowski replied:
Done

salarenko commented on src/input-group.scss:
```diff
@@ -201,6 +201,31 @@ $fd-input-border-radius: var(--sapField_BorderCornerRadius);
+        background-color: var(--sapButton_Lite_Hover_Background);
+
+        &.is-expanded {
+          background-color: var(--sapButton_Active_Background);
```
Shouldn't this be refactored to use SCSS variables initialized with CSS variables or values at the top of this selector?  (Same for the rest of edited components)

JKMarkowski replied:
Done

JKMarkowski commented:
> * Disabled Select, Combobox and Multi-input can be opened using keyboard
>   ![image](https://user-images.githubusercontent.com/17496353/73538610-b938a080-442b-11ea-8947-df877127d8a5.png)

I don't think if there should be anything preventing from opening popover in such a case.

> * Combobox and Multi-input input button doesn't take full height
>   ![image](https://user-images.githubusercontent.com/17496353/73538721-01f05980-442c-11ea-8409-6deff9429946.png)

It happens only, when you zoom in/out

> * Missing class that makes the input wide (also disabled example)
>   ![image](https://user-images.githubusercontent.com/17496353/73538820-44199b00-442c-11ea-8c82-eda2aea701b2.png)

It's fixed now, there is no wide class anywhere anymore

>   ![image](https://user-images.githubusercontent.com/17496353/73539482-c48ccb80-442d-11ea-97ac-593831aee83f.png)
> * Please describe classes used for setting components in specified visual state. Is there no for example `is-disabled`? We should only base on `aria`?

:+1:

InnaAtanasova commented:
<img width="322" alt="Screen Shot 2020-02-03 at 4 55 18 PM" src="https://user-images.githubusercontent.com/39598672/73694416-f90abc80-46a5-11ea-9fab-22269a86ce48.png">

The content of the "body" is cutting the border-radiuses of the popover

InnaAtanasova commented:
<img width="139" alt="Screen Shot 2020-02-03 at 4 54 42 PM" src="https://user-images.githubusercontent.com/39598672/73694457-0f187d00-46a6-11ea-9206-2b86baeef788.png">
The outline should be white?

InnaAtanasova commented:
<img width="507" alt="Screen Shot 2020-02-03 at 4 58 49 PM" src="https://user-images.githubusercontent.com/39598672/73694647-7e8e6c80-46a6-11ea-9781-1076377815d0.png">

The dropdown body should have border-radius only at the bottom.

JKMarkowski commented:
> I'm not seeing the bottom border on the selected item for this theme in the specs
> 
> <img alt="Screen Shot 2020-02-03 at 5 27 18 PM" width="340" src="https://user-images.githubusercontent.com/2471874/73702596-76d0c700-46aa-11ea-9d46-27a9829749e6.png">

We had a talk with Inna about it and it should be there.
Combobox and multi input should share body with select/list

droshev commented:
@estelaxu3 could you take a look on this component?

estelaxu3 commented:
In multi input, the icon is styles is not matching Fiori specs:
Styles:
<img width="699" alt="Screen Shot 2020-02-06 at 10 06 02 AM" src="https://user-images.githubusercontent.com/54589269/73949392-5d0ccb00-48c8-11ea-9af9-de40ffc6d5bc.png">

Fiori3:
<img width="298" alt="Screen Shot 2020-02-06 at 10 06 21 AM" src="https://user-images.githubusercontent.com/54589269/73949496-7b72c680-48c8-11ea-89e3-56c746b11787.png">

estelaxu3 commented:
Also, I noticed that in Search input, the visuals is not completely matched. The first image is from Fiori3 and second is from Styles.
<img width="753" alt="Screen Shot 2020-02-06 at 10 21 48 AM" src="https://user-images.githubusercontent.com/54589269/73950813-90e8f000-48ca-11ea-95d6-6027eb61f8dd.png">

JKMarkowski commented:
@estelaxu3 search input is deprecated on our site, icon will be change in example.

estelaxu3 commented:
The rest looks good to me.

JKMarkowski commented:
> In multi input, the icon on the right corner in Styles is not matching Fiori specs:
> Styles:
> <img alt="Screen Shot 2020-02-06 at 10 06 02 AM" width="699" src="https://user-images.githubusercontent.com/54589269/73949392-5d0ccb00-48c8-11ea-9af9-de40ffc6d5bc.png">
> 
> Fiori3:
> <img alt="Screen Shot 2020-02-06 at 10 06 21 AM" width="298" src="https://user-images.githubusercontent.com/54589269/73949496-7b72c680-48c8-11ea-89e3-56c746b11787.png">

![image](https://user-images.githubusercontent.com/26483208/73955719-13da6c80-4904-11ea-954c-fb89ada77da6.png)

it's from docs, I used good icon, some of icons in docs are different than others

droshev commented:
@JKMarkowski could you list all the class names that were removed with this PR?

JKMarkowski commented:
> @JKMarkowski could you list all the class names that were removed with this PR?
> @Vanessa-Cusmich Could you review the documentation wording?
> 
> I think it is important to write in which case which component should be used (e.g. for less than 12 items then the developer should use `select` for more than 12 items the developer can use `combobox`, etc.)
> 
> Here are my comments regarding:
> `Select`
> 
> * can we make the documentation of the example a little bit more verbose(e.g. explaining what 2 columns is, etc.) The idea is to add some description what the examples are for. You did that for Select text wrapping.

Added 

> * Can you add an example of select with text and icon in the menu?

Added

> * I am not sure whether the button of the select should get the focus when it is clicked. I couldn't find this on the documentation. The focus there is on the entire field including the button.

I'm not sure either, it's about adding `tabindex='-1'` argument to example

> 
> <img alt="Screen Shot 2020-02-09 at 11 14 34 AM" width="135" src="https://user-images.githubusercontent.com/4380815/74105627-811d1600-4b2d-11ea-8d27-2ccca9d8b170.png">
> 
> * (Question) Should we open a separate issue for select rendered on mobile?

I think we do it in separated issue

> * (Question) Does the select more like a pattern in our case?

I would say select is component, because it has it's own styles.

> 
> `Combobox`
> 
> * I think it is good to write what the component is about (e.g. `combobox` allows users to select an item from a predefined list and it provides an editable input field for filtering the list, and a dropdown menu with a list of the available options. If the entries are not validated by the application, users can also enter custom values
> * It is hard to understand:
> 
> ```
> The combobox input component is an opinionated composition of the input group, popover and list components. The popover is shared between multi-input and select component. In Select there are much more examples of modifications that can be provided also to combobox component.
> ```
> 
> Also the combobox is a pattern and not a component
> 
> * Can you change the first example the field and menu to be with the same width?
> * (Question) Should we open a separate issue for select rendered on mobile?
> * I am not sure whether the button of the combobox should get the focus when it is clicked. I couldn't find this on the documentation. The focus there is on the entire field including the button.
> * Can you add examples with Two-Column Layout and Grouping?
> 
> `Multi Input`
> 
> * I think the description here is a copy from the combobox:
> 
> ```
> The combobox input component is an opinionated composition of the input group, popover and list components. The popover is shared between multi-input and select component. In Select there are much more examples of modifications that can be provided also to combobox component.
> ```
> 
> Can you update it and see my comment for `combobox`?
> 
> * Can you make at least the first examples the token data to match the menu underneath?
> 
> <img alt="Screen Shot 2020-02-09 at 11 54 24 AM" width="609" src="https://user-images.githubusercontent.com/4380815/74106245-00611880-4b33-11ea-8d19-bb46375a89e3.png">
> 
> * Can we add a description what the pattern is about (e.g. A multi-input field allows the user to enter multiple values, which are displayed as tokens. To help the user enter a valid value, you can enable the suggestions feature and the value help option.)
> * (Question) Should we open a separate issue for select rendered on mobile?
> * Can you add an example with Grouping?

JKMarkowski commented:
hi @droshev I think I added all of your hints. Please take a look on it and we can merge it

droshev commented:
@JKMarkowski let me know once you apply the comments that were discussed during the code review

JKMarkowski commented:
Hello @droshev  changes are provided anr PR is ready to check
